src/pip/_internal/exceptions.py

```diff
@@ -30,34 +143,48 @@ class UninstallationError(PipError):
     """General exception during uninstallation"""
 
 
-class NoneMetadataError(PipError):
-    """
-    Raised when accessing "METADATA" or "PKG-INFO" metadata for a
-    pip._vendor.pkg_resources.Distribution object and
-    `dist.has_metadata('METADATA')` returns True but
-    `dist.get_metadata('METADATA')` returns None (and similarly for
-    "PKG-INFO").
+class MetadataFileIsMissingError(DiagnosticPipError, InstallationError):
+    """Raised when a distribution does not have a "METADATA" or "PKG-INFO" file."""
+
+    reference = "metadata-file-missing"
+
+    def __init__(self, dist: Distribution, error: Exception) -> None:
+        context = [str(error)]
+        if dist.location is not None:
+            context.append(f"This distribution is located at\n{dist.location}")
+
+        super().__init__(
+            message=f"Could not find metadata file for {dist}",
+            context="\n".join(context),
+            hint="This is likely due to an improperly generated package.",
```

Comment:
Nit: one thing that I never know is whether to write a dot at the end of the message or not. Would this be an opportunity to enforce consistency ?

Reply:
These are full sentences, always. :)

Reply:
Didn't enforce this for message / context, but added an `_stmt` to the other two + added an assertion to ensure this is valid always. :)